File: sdk-typescript/src/governance/strict-receipt-recovery-v2.ts
import { createHash, createPublicKey, verify as cryptoVerify } from 'node:crypto';
import type { IntentAlignmentV2Result, IntentV2BaseResult } from '../policy/intent-alignment-v2.js';
import { canonicalJsonForHash } from '../policy/tool-pinning.js';
import type { DeviceSigner } from '../proxy/device-identity.js';
import type { ActionContextV2Document, PriorActionV2Input } from './action-context-v2.js';
import type { StrictReceiptV2Envelope } from './strict-receipt-v2.js';
import { strictReceiptV2KeyId } from './strict-receipt-v2.js';

export const STRICT_RECOVERY_V2_SCHEMA = 'obsvr-strict-receipt-recovery-v2' as const;
export const STRICT_RECOVERY_V2_ENVELOPE_SCHEMA = 'obsvr-strict-receipt-recovery-envelope-v2' as const;
const DOMAIN = Buffer.from('obsvr-strict-receipt-recovery/2\0', 'utf8');
const SPKI = Buffer.from('302a300506032b6570032100', 'hex');
const HEX64 = /^[0-9a-f]{64}$/;
const HEX128 = /^[0-9a-f]{128}$/;

export interface RecoveryPendingV2 {
  kind: 'decision' | 'resolution' | 'timeout';
  receipt: StrictReceiptV2Envelope;
  context?: ActionContextV2Document;
  base_result?: IntentV2BaseResult;
  evaluation?: IntentAlignmentV2Result;
  suspended_receipt_hash?: string;
}
export interface RecoveryCommittedV2 {
  sequence: number;
  head_receipt_hash: string | null;
  last_timestamp_ms: number | null;
  prior_actions: PriorActionV2Input[];
  suspended: Array<{ receipt_hash: string; receipt: StrictReceiptV2Envelope;
    context: ActionContextV2Document; base_result: IntentV2BaseResult }>;
  resolved_receipt_hashes: string[];
  action_ids: string[];
  approval_requests: Array<{ request_id: string; receipt_hash: string }>;
}
export interface StrictRecoveryV2Document {
  schema: typeof STRICT_RECOVERY_V2_SCHEMA;
  profile_version: '2.0';
  tenant_id: string;
  session_id: string;
  sdk_language: 'typescript';
  sdk_version: string;
  origin_pid: number;
  committed: RecoveryCommittedV2;
  prepared?: RecoveryPendingV2;
}
export interface StrictRecoveryV2Envelope {
  schema: typeof STRICT_RECOVERY_V2_ENVELOPE_SCHEMA;
  document: StrictRecoveryV2Document;
  checkpoint_hash: string;
  signature: { algorithm: 'Ed25519'; key_id: string; value: string };
}

export class StrictRecoveryV2Error extends Error {
  constructor(message: string) { super(message); this.name = 'StrictRecoveryV2Error'; }
}
function fail(message: string): never { throw new StrictRecoveryV2Error(message); }
function text(value: unknown, field: string): string {
  if (typeof value !== 'string' || value.trim().length === 0) return fail(`${field} must be nonblank`);
  return value;
}
function integer(value: unknown, field: string): number {
  if (typeof value !== 'number' || !Number.isSafeInteger(value) || value < 0) return fail(`${field} must be a nonnegative safe integer`);
  return value;
}
function checkpointHash(document: StrictRecoveryV2Document): string {
  return createHash('sha256').update(canonicalJsonForHash(document)).digest('hex');
}
function preimage(hash: string): Buffer { return Buffer.concat([DOMAIN, Buffer.from(hash, 'hex')]); }
function verifySignature(envelope: StrictRecoveryV2Envelope, rawPublicKey: Uint8Array): boolean {
  const key = createPublicKey({ key: Buffer.concat([SPKI, Buffer.from(rawPublicKey)]), format: 'der', type: 'spki' });
  return cryptoVerify(null, preimage(envelope.checkpoint_hash), key, Buffer.from(envelope.signature.value, 'hex'));
}

export function signStrictRecoveryV2(
  document: StrictRecoveryV2Document, signer: DeviceSigner,
): StrictRecoveryV2Envelope {
  validateStrictRecoveryV2Document(document);
  const checkpoint_hash = checkpointHash(document);
  const keyId = strictReceiptV2KeyId(signer.rawPublicKey);
  const value = signer.signBytes(preimage(checkpoint_hash));
  if (!HEX128.test(value)) fail('checkpoint signer returned an invalid signature');
  const envelope: StrictRecoveryV2Envelope = { schema: STRICT_RECOVERY_V2_ENVELOPE_SCHEMA,
    document: JSON.parse(JSON.stringify(document)) as StrictRecoveryV2Document,
    checkpoint_hash, signature: { algorithm: 'Ed25519', key_id: keyId, value } };
  if (!verifySignature(envelope, signer.rawPublicKey)) fail('checkpoint signature failed self-verification');
  return envelope;
}

export function verifyStrictRecoveryV2(
  value: unknown, signer: DeviceSigner,
): StrictRecoveryV2Document {
  if (!value || typeof value !== 'object') return fail('checkpoint must be an object');
  const envelope = value as StrictRecoveryV2Envelope;
  if (envelope.schema !== STRICT_RECOVERY_V2_ENVELOPE_SCHEMA
    || !HEX64.test(envelope.checkpoint_hash)
    || envelope.signature?.algorithm !== 'Ed25519'
    || envelope.signature.key_id !== strictReceiptV2KeyId(signer.rawPublicKey)
    || !HEX128.test(envelope.signature.value)) return fail('invalid checkpoint envelope');
  validateStrictRecoveryV2Document(envelope.document);
  if (checkpointHash(envelope.document) !== envelope.checkpoint_hash) return fail('checkpoint hash mismatch');
  if (!verifySignature(envelope, signer.rawPublicKey)) return fail('checkpoint signature invalid');
  return JSON.parse(JSON.stringify(envelope.document)) as StrictRecoveryV2Document;
}

export function validateStrictRecoveryV2Document(document: StrictRecoveryV2Document): void {
  if (!document || typeof document !== 'object' || document.schema !== STRICT_RECOVERY_V2_SCHEMA
    || document.profile_version !== '2.0' || document.sdk_language !== 'typescript') return fail('invalid checkpoint document');
  text(document.tenant_id, 'tenant_id'); text(document.session_id, 'session_id'); text(document.sdk_version, 'sdk_version');
  integer(document.origin_pid, 'origin_pid'); integer(document.committed?.sequence, 'committed.sequence');
  if ((document.committed.head_receipt_hash === null) !== (document.committed.sequence === 0)) return fail('checkpoint head/sequence mismatch');
  if (document.committed.head_receipt_hash !== null && !HEX64.test(document.committed.head_receipt_hash)) return fail('invalid checkpoint head hash');
  if (document.committed.last_timestamp_ms !== null) integer(document.committed.last_timestamp_ms, 'last_timestamp_ms');
  if (!Array.isArray(document.committed.prior_actions) || !Array.isArray(document.committed.suspended)
    || !Array.isArray(document.committed.resolved_receipt_hashes) || !Array.isArray(document.committed.action_ids)
    || !Array.isArray(document.committed.approval_requests)) return fail('checkpoint collections must be arrays');
  const last = document.committed.prior_actions.at(-1);
  if (document.committed.sequence > 0 && (!last || last.sequence !== document.committed.sequence
    || last.receipt_hash !== document.committed.head_receipt_hash)) return fail('checkpoint head does not match prior actions');
  if (document.prepared) {
    if (!['decision', 'resolution', 'timeout'].includes(document.prepared.kind)
      || document.prepared.receipt.body.tenant_id !== document.tenant_id
      || document.prepared.receipt.body.session_id !== document.session_id
      || document.prepared.receipt.body.profile_version !== '2.0'
      || document.prepared.receipt.body.sequence !== document.committed.sequence + 1
      || document.prepared.receipt.body.previous_receipt_hash !== document.committed.head_receipt_hash) return fail('prepared receipt does not continue checkpoint head');
  }
}

File: sdk-python/obsvr/strict_receipt_recovery_v2.py
"""Authenticated durable checkpoints for strict v2 coordinator recovery."""

from __future__ import annotations

import copy
import hashlib
from typing import Any, Dict

from .strict_receipt_v2 import strict_receipt_v2_key_id
from .tool_pinning import _canonical_json_for_hash

STRICT_RECOVERY_V2_SCHEMA = "obsvr-strict-receipt-recovery-v2"
STRICT_RECOVERY_V2_ENVELOPE_SCHEMA = "obsvr-strict-receipt-recovery-envelope-v2"
_DOMAIN = b"obsvr-strict-receipt-recovery/2\x00"
_HEX = frozenset("0123456789abcdef")
_MAX_SAFE = 9_007_199_254_740_991


class StrictRecoveryV2Error(ValueError):
    """A recovery checkpoint is malformed, untrusted, or inconsistent."""


def _text(value: Any, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise StrictRecoveryV2Error(f"{field} must be nonblank")
    return value


def _integer(value: Any, field: str) -> int:
    if (
        isinstance(value, bool)
        or not isinstance(value, int)
        or value < 0
        or value > _MAX_SAFE
    ):
        raise StrictRecoveryV2Error(f"{field} must be a nonnegative safe integer")
    return value


def _hash(value: Any, field: str) -> str:
    if (
        not isinstance(value, str)
        or len(value) != 64
        or any(char not in _HEX for char in value)
    ):
        raise StrictRecoveryV2Error(f"{field} must be 64 lowercase hex characters")
    return value


def _checkpoint_hash(document: Dict[str, Any]) -> str:
    return hashlib.sha256(
        _canonical_json_for_hash(document).encode("utf-8")
    ).hexdigest()


def _preimage(checkpoint_hash: str) -> bytes:
    return _DOMAIN + bytes.fromhex(checkpoint_hash)


def _verify(raw_public_key: bytes, message: bytes, signature: bytes) -> bool:
    try:
        from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey

        Ed25519PublicKey.from_public_bytes(raw_public_key).verify(signature, message)
        return True
    except Exception:
        try:
            from nacl.signing import VerifyKey

            VerifyKey(raw_public_key).verify(message, signature)
            return True
        except Exception:
            return False


def validate_strict_recovery_v2_document(document: Dict[str, Any]) -> None:
    if (
        not isinstance(document, dict)
        or document.get("schema") != STRICT_RECOVERY_V2_SCHEMA
        or document.get("profile_version") != "2.0"
        or document.get("sdk_language") != "python"
        or not isinstance(document.get("committed"), dict)
    ):
        raise StrictRecoveryV2Error("invalid checkpoint document")
    _text(document.get("tenant_id"), "tenant_id")
    _text(document.get("session_id"), "session_id")
    _text(document.get("sdk_version"), "sdk_version")
    _integer(document.get("origin_pid"), "origin_pid")
    committed = document["committed"]
    sequence = _integer(committed.get("sequence"), "committed.sequence")
    head = committed.get("head_receipt_hash")
    if (head is None) != (sequence == 0):
        raise StrictRecoveryV2Error("checkpoint head/sequence mismatch")
    if head is not None:
        _hash(head, "head_receipt_hash")
    timestamp = committed.get("last_timestamp_ms")
    if timestamp is not None:
        _integer(timestamp, "last_timestamp_ms")
    for field in (
        "prior_actions",
        "suspended",
        "resolved_receipt_hashes",
        "action_ids",
        "approval_requests",
    ):
        if not isinstance(committed.get(field), list):
            raise StrictRecoveryV2Error("checkpoint collections must be arrays")
    prior = committed["prior_actions"]
    if sequence and (
        not prior
        or prior[-1].get("sequence") != sequence
        or prior[-1].get("receipt_hash") != head
    ):
        raise StrictRecoveryV2Error("checkpoint head does not match prior actions")
    prepared = document.get("prepared")
    if prepared is not None:
        if not isinstance(prepared, dict) or prepared.get("kind") not in (
            "decision",
            "resolution",
            "timeout",
        ):
            raise StrictRecoveryV2Error("invalid prepared checkpoint")
        receipt = prepared.get("receipt", {})
        body = receipt.get("body", {}) if isinstance(receipt, dict) else {}
        if (
            body.get("tenant_id") != document["tenant_id"]
            or body.get("session_id") != document["session_id"]
            or body.get("profile_version") != "2.0"
            or body.get("sequence") != sequence + 1
            or body.get("previous_receipt_hash") != head
        ):
            raise StrictRecoveryV2Error(
                "prepared receipt does not continue checkpoint head"
            )


def sign_strict_recovery_v2(document: Dict[str, Any], signer: Any) -> Dict[str, Any]:
    validate_strict_recovery_v2_document(document)
    checkpoint_hash = _checkpoint_hash(document)
    key_id = strict_receipt_v2_key_id(signer.raw_public_key)
    value = signer.sign_bytes(_preimage(checkpoint_hash))
    if len(value) != 128 or any(char not in _HEX for char in value):
        raise StrictRecoveryV2Error("checkpoint signer returned an invalid signature")
    if not _verify(
        signer.raw_public_key, _preimage(checkpoint_hash), bytes.fromhex(value)
    ):
        raise StrictRecoveryV2Error("checkpoint signature failed self-verification")
    return {
        "schema": STRICT_RECOVERY_V2_ENVELOPE_SCHEMA,
        "document": copy.deepcopy(document),
        "checkpoint_hash": checkpoint_hash,
        "signature": {"algorithm": "Ed25519", "key_id": key_id, "value": value},
    }


def verify_strict_recovery_v2(value: Any, signer: Any) -> Dict[str, Any]:
    if not isinstance(value, dict):
        raise StrictRecoveryV2Error("checkpoint must be an object")
    signature = value.get("signature")
    if (
        value.get("schema") != STRICT_RECOVERY_V2_ENVELOPE_SCHEMA
        or not isinstance(signature, dict)
        or signature.get("algorithm") != "Ed25519"
        or signature.get("key_id") != strict_receipt_v2_key_id(signer.raw_public_key)
    ):
        raise StrictRecoveryV2Error("invalid checkpoint envelope")
    checkpoint_hash = _hash(value.get("checkpoint_hash"), "checkpoint_hash")
    signature_hex = signature.get("value")
    if (
        not isinstance(signature_hex, str)
        or len(signature_hex) != 128
        or any(char not in _HEX for char in signature_hex)
    ):
        raise StrictRecoveryV2Error("invalid checkpoint envelope")
    document = value.get("document")
    validate_strict_recovery_v2_document(document)
    if _checkpoint_hash(document) != checkpoint_hash:
        raise StrictRecoveryV2Error("checkpoint hash mismatch")
    if not _verify(
        signer.raw_public_key,
        _preimage(checkpoint_hash),
        bytes.fromhex(signature_hex),
    ):
        raise StrictRecoveryV2Error("checkpoint signature invalid")
    return copy.deepcopy(document)
